### PR TITLE
feat: role based views

### LINKED ISSUE #281: feat: add filtering logic (roles) to dm-core that can be used by ui plugin selector
The goal is to have different views for different roles. 

When a user log in to the web app, he is assigned a role. The role for users are defined in azure: https://portal.azure.com/?feature.msaljs=true#view/Microsoft_AAD_RegisteredApps/ApplicationMenuBlade/~/AppRoles/appId/97a6b5bd-63fb-42c6-bb75-7e5de2394ba0/isMSAApp~/false

you can assign a role to a user inside the enterprise application "Marine Analyis Framework Data Modelling Tool "  - https://portal.azure.com/#view/Microsoft_AAD_IAM/ManagedAppMenuBlade/~/Users/objectId/b9041025-05f0-44d4-89a7-3b5f955c0de5/appId/97a6b5bd-63fb-42c6-bb75-7e5de2394ba0

The roles concept has been implemented before in the analysis platform app. In this app, we added roles to the UI plugin like this: 

Image

the roles list defines all roles that should display a given plugin. So the inspector role should display the view-analysis-inspector plugin and other roles should display a different plugin

note: the userInfo component, used in the Header plugin, can be used to view your role.

Image

----------------------
```
{
  "type": "CORE:RecipeLink",
  "_blueprintPath_": "/plugins/grid/blueprints/Dashboard",
  "initialUiRecipe": {
    "name": "Dashboard",
    "type": "CORE:UiRecipe",
    "roles": ["Operator", "Admin"],
    "plugin": "@development-framework/dm-core-plugins/grid",
    "config": {
      "type": "PLUGINS:dm-core-plugins/grid/GridPluginConfig",
      "size": {
        "type": "PLUGINS:dm-core-plugins/grid/GridSize",
        "columns": 6,
        "rows": 5
      },
      "items": [
        {
          "type": "PLUGINS:dm-core-plugins/grid/GridItem",
          "viewConfig": {
            "type": "CORE:ReferenceViewConfig",
            "scope": "aNestedObjectWithCustomUI",
            "recipe": "aRecipeName"
          },
          "gridArea": {
            "type": "PLUGINS:dm-core-plugins/grid/GridArea",
            "rowStart": 1,
            "columnStart": 1,
            "rowEnd": 3,
            "columnEnd": 6
          }
        },
        {
          "type": "PLUGINS:dm-core-plugins/grid/GridItem",
          "viewConfig": {
            "type": "CORE:InlineRecipeViewConfig",
            "scope": "aNestedObjectWithCustomUI",
            "recipe": {
              "type": "CORE:UiRecipe",
              "plugin": "@development-framework/dm-core-plugins/form",
              "name": "form",
              "config": {
                "type": "PLUGINS:dm-core-plugins/form/FormInput",
                "attributes": [
                  {
                    "type": "PLUGINS:dm-core-plugins/form/fields/StringField",
                    "name": "bar",
                    "widget": "TextareaWidget"
                  }
                ]
              }
            }
          },
          "gridArea": {
            "type": "PLUGINS:dm-core-plugins/grid/GridArea",
            "rowStart": 1,
            "columnStart": 4,
            "rowEnd": 3,
            "columnEnd": 6
          }
        },
        {
          "type": "PLUGINS:dm-core-plugins/grid/GridItem",
          "viewConfig": {
            "type": "CORE:ReferenceViewConfig",
            "scope": "orders[0].product",
            "recipe": "Yaml"
          },
          "gridArea": {
            "type": "PLUGINS:dm-core-plugins/grid/GridArea",
            "rowStart": 5,
            "columnStart": 3,
            "rowEnd": 6,
            "columnEnd": 7
          }
        },
        {
          "type": "PLUGINS:dm-core-plugins/grid/GridItem",
          "viewConfig": {
            "type": "CORE:ReferenceViewConfig",
            "scope": "order.product",
            "recipe": "Form"
          },
          "gridArea": {
            "type": "PLUGINS:dm-core-plugins/grid/GridArea",
            "rowStart": 3,
            "columnStart": 3,
            "rowEnd": 6,
            "columnEnd": 5
          }
        },
        {
          "type": "PLUGINS:dm-core-plugins/grid/GridItem",
          "viewConfig": {
            "type": "CORE:InlineRecipeViewConfig",
            "recipe": {
              "type": "CORE:UiRecipe",
              "plugin": "@development-framework/dm-core-plugins/json",
              "name": "Json"
            }
          },
          "gridArea": {
            "type": "PLUGINS:dm-core-plugins/grid/GridArea",
            "rowStart": 3,
            "columnStart": 1,
            "rowEnd": 6,
            "columnEnd": 3
          }
        }
      ]
    }
  },
  "uiRecipes": [
    {
      "name": "Yaml",
      "type": "CORE:UiRecipe",
      "plugin": "@development-framework/dm-core-plugins/yaml"
    },
    {
      "name": "Edit",
      "type": "CORE:UiRecipe",
      "plugin": "@development-framework/dm-core-plugins/form"
    }
  ]
}

```

### PR DESCRIPTION
**Pull request by collinlokken**
## What does this pull request change?
Enables role based views in HeaderPlugin and ViewSelectorPlugin by filtering out UiRecipes and Views that do not match the role of the user.

## Why is this pull request needed?
Makes it possible to customise apps based on the user

## Issues related to this change
Closes #281

### CODE REVIEW

File: packages/dm-core/src/utils/arrayIntersection.tsx [hunk 1]
```diff
@@ -0,0 +1,8 @@
+export const arrayIntersection = (
+  array1: string[],
+  array2: string[]
+): string[] => {
+  const result = array1.filter((item) => array2.includes(item))
+  console.log(array1, array2, result)
+  return result
+}
```
Comment (henrikbossart) [inline packages/dm-core/src/utils/arrayIntersection.tsx:8]: I notice that we already have dependencies that use `lodash`, so we might as well use it ourselves. If we do, we don't have to write a lot of helper functions, like this one, and can instead just import them from lodash.

Comment (collinlokken) [reply packages/dm-core/src/utils/arrayIntersection.tsx:8]: Lets try lodash!

File: packages/dm-core-plugins/src/role_filter/RoleFilterPlugin.tsx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  )
+  const [openViewConfigs, setOpenViewConfigs] = useState<filteredView[]>([])
+  const [selectedRole, setSelectedRole] = useState<string>()
+  const [allowedRoles, setAllowedRoles] = useState<string[]>([])
+
+  // TODO: Implement real roles when User login is in place
+  const User = {
+    roles: ['admin', 'operator'],
+  }
+  type filteredView = {
```
Comment (eoaksnes) [inline packages/dm-core-plugins/src/role_filter/RoleFilterPlugin.tsx:19]: The type should be `PascalCase`. Also, maybe move it outside RoleFilterPlugin, just before?

**Review by eoaksnes — APPROVED** [review]
Nice! This will be a great feature!

**Pull request merged**